### PR TITLE
[FLINK-25368][connectors/kafka] Substitute KafkaConsumer with AdminClient when getting offsets

### PR DESCRIPTION
**Pull request by dengziming**
## What is the purpose of the change
`AdminClient.listOffsets` is provided in Kafka 2.7, In the future more `OffsetSpce` types will be added to it, for example, OffsetSpec.MaxTimestampSpec is added in Kafka 3.0.0. so it's better to substitute `KafkaConsumer` with `AdminClient`.

## Brief change log

  - Substitute KafkaConsumer with AdminClient in OffsetsInitializer
  - Substitute KafkaConsumer with AdminClient in FlinkKafkaConsumer
  - Totally remove KafkaConsumer from KafkaSourceEnumerator.

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change is already covered by existing tests `OffsetsInitializerTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation
  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by dengziming** [comment]
hello @becketqin @fapaul , PTAL.

**Comment by dengziming** [comment]
Also ping @lindong28 @PatrickRen .

**Comment by lindong28** [comment]
Thanks for the creating the ticket to improve Flink!

I think the overall motivation makes sense. But the PR changed OffsetsInitializer::PartitionOffsetsRetriever and OffsetsInitializer is marked @PublicEvolving. We probably need to file a FLIP for any public API change like this.

This PR focuses on refactoring the existing class without providing extra functionality directly. my gut feel is that we can do this refactor only when we really need the extra functionality. Then we will need to describe the motivation and the scenario in the FLIP.

**Comment by dengziming** [comment]
Thank you for your comment @lindong28 , I think your suggestions make sense, we'd better make this change in a related FLIP.

**Comment by fapaul** [comment]
I agree with @lindong28 if we want to change the `OffsetsInitializer#PartitionOffsetsRetriever` we need a FLIP. In this case, there might be also a different alternative. We could introduce a second internal interface extending the existing one and implement the `list` functionality. AFAICT the list function is only used internally by the enumerator. WDYT?

**Comment by dengziming** [comment]
Thank you @fapaul for your suggestions, I think your proposal is viable here and I will try it soon.

**Comment by dengziming** [comment]
I rewrote this PR and tried to keep the `OffsetsInitializer#PartitionOffsetsRetriever` unchanged according to your suggestion, PTAL, @fapaul 🤝

**Review by alpreu — CHANGES_REQUESTED** [review]
Thank you for reworking the PR, looks pretty good to me. I found some minor issues where I left comments

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 1]
```diff
@@ -281,27 +283,50 @@ protected AbstractPartitionDiscoverer createPartitionDiscoverer(
...
+            for (Map.Entry<TopicPartition, Long> partitionToOffset :
+                    topicPartitionOffsets.entrySet()) {
                 result.put(
                         new KafkaTopicPartition(
                                 partitionToOffset.getKey().topic(),
                                 partitionToOffset.getKey().partition()),
                         (partitionToOffset.getValue() == null)
                                 ? null
-                                : partitionToOffset.getValue().offset());
+                                : partitionToOffset.getValue());
```
[the remark this answers is not in the document]
Comment (dengziming) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Good catch!

**Comment by dengziming** [comment]
Thank you @alpreu for your suggestions, done! PTAL again.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 2]
```diff
@@ -281,29 +283,41 @@ protected AbstractPartitionDiscoverer createPartitionDiscoverer(
...
-
-                result.put(
-                        new KafkaTopicPartition(
-                                partitionToOffset.getKey().topic(),
-                                partitionToOffset.getKey().partition()),
-                        (partitionToOffset.getValue() == null)
-                                ? null
-                                : partitionToOffset.getValue().offset());
-            }
+        try (Admin adminClient = Admin.create(properties)) {
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java:295]: I wouldn't add this change to the `FlinkKafkaConsumer` because it is already marked as deprecated and will be removed in one of the next releases. We want to incentivize users to upgrade to the KafkaSource and do not continue using this class.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java [hunk 3]
```diff
@@ -152,7 +152,7 @@ public static void startClusters(boolean secureMode) throws Exception {
 
     public static void startClusters(KafkaTestEnvironment.Config environmentConfig)
             throws Exception {
-        kafkaServer = constructKafkaTestEnvionment();
+        kafkaServer = constructKafkaTestEnvironment();
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java:155]: Nit: Can you make the renaming on a separate commit as hotfix?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java [hunk 4]
```diff
@@ -558,25 +533,100 @@ public PartitionOffsetsRetrieverImpl(
...
+                                + topicPartitionOffsets
+                                + " due to",
+                        e);
+            }
+        }
+
         @Override
         public Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions) {
-            return consumer.endOffsets(partitions);
+            return listOffsets(
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:580]: The methods `endOffset` and `beginningOffsets` look very similar, please extract a common method that takes as input the `OffsetSpec`.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java [hunk 5]
```diff
@@ -558,25 +533,100 @@ public PartitionOffsetsRetrieverImpl(
...
+                                            (tp, listOffsetsResultInfo) -> {
+                                                if (listOffsetsResultInfo != null) {
+                                                    offsets.put(tp, listOffsetsResultInfo);
+                                                }
+                                            });
+                                    return offsets;
+                                })
+                        .get();
+            } catch (InterruptedException e) {
+                throw new FlinkRuntimeException(
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:565]: Please also set the interrupted flag if you handle the `InterruptedException`

```java
Thread.currentThread().interrupt();
```

**Review by fapaul — COMMENTED** [review]
@dengziming thanks for your effort I left a couple of comments. The overall approach looks fine but I think it requires some clean-up.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java (hunk 2 above)
Comment (dengziming) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Yes, I reverted this change since it is already marked as deprecated.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java (hunk 5 above)
Comment (dengziming) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:567]: Thank you for this reminder, I also find there is another `InterruptedException` ad `committedOffsets` so I revised it too.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java (hunk 3 above)
Comment (dengziming) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java:155]: I already rebased my commits and made this change in a separate commit as other PRs, do you mean I should submit another PR for this hotfix?

**Review by fapaul — APPROVED** [review]
LGTM now % please rephrase the last commit message to be more declarative

**Comment by dengziming** [comment]
Thanks for your patience @fapaul , Done! 🤝
